Composite refs

mmussomele opened the pull request:
Fixes #239

mmussomele commented:
Fixes #239

mmussomele commented on ast/check.go:
```diff
@@ -534,6 +538,19 @@ func (rc *refChecker) checkRefLeaf(tpe types.Type, ref Ref, idx int) *Error {
+			return nil
+		}
+
+		unify1(rc.env, SetTerm(head), exist, false)
```
ah yea that would be much better

mmussomele commented on ast/check.go:
```diff
@@ -487,6 +487,9 @@ func (rc *refChecker) checkRef(curr *TypeEnv, node *typeTreeNode, ref Ref, idx i
 		}
 
+	case Array, Object, *Set:
+		rc.env.tree.PutOne(value, rc.env.Get(value))
```
Yea I think you're right, this case just shouldn't exist. I'll just let it hit default

mmussomele commented:
All the issues should be fixed now.

tsandall commented:
This is close, but I think more testing is needed. Specifically, anywhere that was branching based on ref operand type needs to be reviewed. E.g., what happens if you write...

```
p[[x,y]] { ... }
```
And then query `p[[foo.bar.baz[_], "qux"]]`?

I imagine we need to have special handling for nested refs?

What about comprehensions nested inside the composites?

mmussomele commented:
Ready for another look, fixed up up some checking issues and corrected eval to handle refs nested within composite operands. Comprehensions also work inside the composites.

tsandall commented on ast/check.go:
```diff
@@ -534,6 +535,13 @@ func (rc *refChecker) checkRefLeaf(tpe types.Type, ref Ref, idx int) *Error {
+		if !unifies(tpe, types.NewSet(types.A)) {
+			return newRefErrInvalid(ref[0].Location, ref, idx, tpe, types.NewSet(types.A), nil)
+		}
+		unify1(rc.env, head, keys, false)
```
This should check if the head unified w/ the keys.

```
if !unify1(...) { // type error }
```

tsandall commented on ast/check_test.go:
```diff
@@ -738,6 +743,14 @@ func TestCheckRefErrInvalid(t *testing.T) {
 			oneOf: []Value{String("p"), String("q")},
 		},
+		{
+			note:  "composite ref into non-set",
```
What about type errors with terms embedded within composite operands? This is related to the comment from above.

For example...

```
p = {[1],[2],[3]}
```

And then...`p[["foo"]]` or `p[{"a": b"}]`.

mmussomele commented:
Comments addressed.

tsandall commented on topdown/topdown.go:
```diff
@@ -1823,14 +1831,30 @@ func evalRefRuleResultRecSet(t *Topdown, set *ast.Set, ref, path ast.Ref, iter f
 			t.Unbind(undo)
 		}
 		return nil
+	case ast.Array, ast.Object, *ast.Set:
```
This entire function can be simplified to unify the set elements with the next element in the ref. Something like...

```
	head, tail := ref[0], ref[1:]
	k := head.Value
	for _, e := range *set {
		undo, err := evalEqUnify(t, e.Value, k, nil, func(t *Topdown) error {
			return evalRefRuleResultRec(t, PlugValue(k, t.Binding), tail, path.Append(e), iter)
		})
		if err != nil {
			return err
		}
		t.Unbind(undo)
	}
	return nil
```

mmussomele replied:
0.o it's like magic

mmussomele replied:
done

mmussomele commented:
Done
